Miscellaneous fixes mainly for datatypes

pavanbalaji opened the pull request:
Some miscellaneous fixes in preparation for issue #1985.

Depends on #3544.

pavanbalaji commented:
test:jenkins/ch3/most
test:jenkins/ch4/most

pavanbalaji commented:
test:jenkins/ch3/most
test:jenkins/ch4/most

pavanbalaji commented:
test:jenkins/ch3/most
test:jenkins/ch4/most

pavanbalaji commented:
Wesley's warning squash branch (#3544) has been merged into `master` now, so I've rebased this branch on top of `master`.  Not retesting as nothing has changed in this branch (including the commits on which it was based on top of).

pavanbalaji commented on src/include/mpir_dataloop.h:
```diff
@@ -114,12 +114,6 @@
- */
-#define DLOOP_DATALOOP_DEFAULT       0
-#define DLOOP_DATALOOP_ALL_BYTES     1
-
```
I've updated the commit message.  Take a look if this one's better.

pavanbalaji commented on src/mpi/datatype/dataloop/dataloop.c:
```diff
@@ -11,7 +11,6 @@
 #include <stdlib.h>
 
-#undef DEBUG_DLOOP_SIZE
 #undef DLOOP_DEBUG_MEMORY
```
I don't think that's necessary.   This commit removes unused variables.  `DEBUG_DLOOP_SIZE` is unused, it's just not a variable.  I've updated the commit message.

pavanbalaji commented on src/include/mpir_dataloop.h:
```diff
@@ -23,19 +23,6 @@
 
-/* These following dataloop-specific types will be used throughout the DLOOP
- * instance:
- */
```
Done.

pavanbalaji commented on test/mpi/datatype/segtest.c:
```diff
@@ -1,210 +0,0 @@
-
-/*
- * Simple segment test, including timing code
- */
```
Done.

pavanbalaji commented on test/mpi/datatype/indexed_misc.c:
```diff
@@ -301,8 +301,7 @@ int indexed_contig_test(void)
- * order to catch a particular optimization path in MPICH's
- * Dataloop_create_indexed routine */
+ * order to catch a particular optimization path in MPICH */
 int indexed_contig_leading_zero_test(void)
```
Yes, in a sense.  It talks about MPICH internals and assumes that dataloops is the only way to implement datatype internals.

pavanbalaji commented on src/mpi/datatype/looputil.c:
```diff
@@ -923,8 +923,8 @@ void MPIR_Type_release_contents(MPI_Datatype type,
+                         MPI_Aint first, MPI_Aint * lastp, MPL_IOV * vectorp, int *lengthp)
 {
     struct MPIR_Segment_piece_params packvec_params;
     MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIR_SEGMENT_PACK_VECTOR);
```
Fixed.

pavanbalaji commented on src/mpi/datatype/dataloop/dataloop_create.c:
```diff
@@ -34,6 +34,13 @@ void MPIR_Dataloop_create(MPI_Datatype type, MPIR_Dataloop ** dlp_p, MPI_Aint *
+        MPIR_Dataloop_create_pairtype(type, dlp_p, dlsz_p);
+        return;
+    }
+
```
The intent was that `MPIR_Dataloop_create` be the public interface for the dataloops code.  `MPIR_Dataloop_create_vector`, `MPIR_Dataloop_create_struct`, `MPIR_Dataloop_create_pairtype`, etc., are supposed to be internal functions that `MPIR_Dataloop_create` would call based on the type of datatype used.  We were using that model correctly for all datatypes, except pairtypes.  This function makes that consistent.

With respect to ARMCI-MPI or OSHMPI, if you are creating derived datatypes frequently in the performance critical path, you have bigger problems than that one extra function-call overhead.  :-)  Compared to the total cost of that function, the extra function call would be in the noise.

minsii replied:
Unlike `vector|struct...`, each pairtype's data loop structure never changes. It somehow makes sense to me that we handle pairtype's data loop separately. I agree that the additional overhead is small, but it reads like we add unnecessary overhead to every creation call just for handling static data loop structures.

For the public interface aspect, I do not have strong preference with either option.

pavanbalaji replied:
@minsii I agree with you in general that pairtypes have some potential for better optimization, but I want to make two points:

1. We don't special case them today.  The code path is the same as other derived datatypes, which is very high overhead.

2. Pair types are not easy to optimize because the MPI standard requires us to not touch the padding bytes, which makes it noncontiguous.  Nathan had a proposal at the MPI Forum to remove the padding byte protection, at which point we might be able to do a fastpath for pairtypes.  But even then the fastpath would essentially make pairtypes as "builtin" types, rather than a derived datatype.  So this change wouldn't matter.

Finally, I want to remind you that here you are arguing for one extra function call and one branch.  That's it.  There's no additional overhead that's being added.  I'd be shocked if you can find a system where this overhead would be noticeable compared with the cost of the `MPIR_Dataloop_create_pairtype` function itself.

For the public interface, I'm not sure what "either option" is here.  The intent was always that `MPIR_Dataloop_create` be the public interface and everything else be a private interface.  Somehow someone patched a hole in that interface (maybe because the namespace was not cleanly done) and called the `MPIR_Dataloop_create_pairtype` function outside of the dataloop code.  There are other such examples too, which I'm slowly fixing.

pavanbalaji replied:
Sorry, I might have misunderstood your concern a bit here.  You are arguing for the cost of `MPIR_Dataloop_create`, not that of `MPIR_Dataloop_create_pairtype`, aren't you?  In that case, the cost is just this one branch that'll never be taken in your example:

```
    if (type == MPI_FLOAT_INT || type == MPI_DOUBLE_INT ||
         type == MPI_LONG_INT || type == MPI_SHORT_INT ||
         type == MPI_LONG_DOUBLE_INT || type == MPI_2INT) {
         MPIR_Dataloop_create_pairtype(type, dlp_p, dlsz_p);
         return;
    }
```

Is that really a concern?  This is during the actual datatype creation, not during communication.  I understand that in ARMCI-MPI, the datatype creation is done in the performance critical path, but the cost of this branch would be negligible even for very small messages.  I find it to be a very weak reason to poke holes in the public interface as an optimization path.

minsii replied:
This condition check is always performed at every `MPIR_Dataloop_create` call. It may take at least 6 compare instructions and 1 branch. Anyway, I agree that the overhead should be small comparing to the other part of this call.

You can add my signoff in this commit.

pavanbalaji replied:
No, that's not true.  Any decent compiler worth its salt will convert that to a labeled jump statement (similar to what you'd get with a switch statement).  So it doesn't depend on the number of checks.

pavanbalaji commented on test/mpi/datatype/indexed_misc.c:
```diff
@@ -347,8 +347,7 @@ int indexed_contig_leading_zero_test(void)
-     * type to force a more complicated datatype processing path
-     * (MPIR_Segment_manipulate in MPICH) */
+     * type to force a more complicated datatype processing path */
     len[0] = 1;
```
Yes.  Merged.

pavanbalaji commented:
Added signoffs and rebased on top of `mpich/master`.  Restarting testing, as the RMA tests should be fixed in `master` now.

test:jenkins/ch3/most
test:jenkins/ch4/most
